### PR TITLE
Loosen restriction to built-in log levels

### PR DESCRIPTION
**Pull request by mgasner**
This will allow users to write their own loggers with extra log levels, e.g. using the verboselogs package.

### CODE REVIEW

File: python_modules/dagster/dagster/core/log.py [hunk 1]
```diff
@@ -81,13 +81,14 @@ def _log(self, method, orig_message, message_props):
...
 
         for logger in self.loggers:
-            logger_method = check.is_callable(getattr(logger, method))
-            if logger.name == DAGSTER_DEFAULT_LOGGER:
-                logger_method(
-                    msg_with_multiline_structured_props, extra={DAGSTER_META_KEY: all_props}
-                )
-            else:
-                logger_method(msg_with_structured_props, extra={DAGSTER_META_KEY: all_props})
+            if hasattr(logger, method):
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/log.py]: I'm concerned about failing silently if we supply the wrong method here

File: python_modules/dagster/dagster/core/log.py [hunk 2]
```diff
@@ -132,3 +133,9 @@ def critical(self, msg, **kwargs):
 
         See debug()'''
         return self._log('critical', msg, kwargs)
+
+    def __getattr__(self, name):
+        def handler(msg, **kwargs):
+            return self._log(name.lower(), msg, kwargs)
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/log.py]: This seems problematic, especially when combined with above issue. what if they are using a library that doesn't user pythonic naming conventions?

File: python_modules/dagster/dagster/core/log.py (hunk 1 above)
Comment (mgasner) [reply python_modules/dagster/dagster/core/log.py]: Yeah, I think this is exposing issues we currently face to some degree but haven't yet considered.

If I configure several loggers (say, a verbose logger or custom logger with my own log levels, and also something out of the box) one of which has the `.notice` method defined, and one of which doesn't, should we fail at runtime when `context.log.notice()` is called, or should we call `.notice` on the loggers that have it defined? (And what should we do if we accidentally call `.notise()`)

Comment (schrockn) [reply python_modules/dagster/dagster/core/log.py]: ah i see. that makes sense. 

I think if the user is using multiple loggers and they have different log levels we should force them to declare what the logging interface they want is at some level. e.g. they should have to say that if they log to log.some_level how that translates to console logging

File: python_modules/dagster/dagster/core/log.py (hunk 2 above)
Comment (mgasner) [reply python_modules/dagster/dagster/core/log.py]: we could omit the call to `.lower()` here -- but all that `._log` does is dispatch to the underlying loggers

File: python_modules/dagster/dagster/core/log.py [hunk 3]
```diff
@@ -80,14 +83,25 @@ def _log(self, method, orig_message, message_props):
...
+                logger_found = True
+                logger_method = check.is_callable(getattr(logger, method))
+                if logger.name == DAGSTER_DEFAULT_LOGGER:
+                    logger_method(
+                        msg_with_multiline_structured_props, extra={DAGSTER_META_KEY: all_props}
+                    )
+                else:
+                    logger_method(msg_with_structured_props, extra={DAGSTER_META_KEY: all_props})
+
+        if not logger_found:
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/log.py]: so the issue for this method is that will be possible to induce crashes at runtime by changing logging config, which I am pretty concerned about.

Comment (mgasner) [reply python_modules/dagster/dagster/core/log.py]: i'm not 100% certain which the cases you're thinking about are --

047d55 reinstates and documents the status quo ante behavior when there are *no* loggers configured (silently passes)

if i have custom loggers which define `.foo()` configured, and call `context.log.foo()` somewhere in my codebase, bnut then remove the custom loggers, the behavior in this diff is that we log an `ERROR` (and continue to execute) -- we could log this as a `WARNING` instead; previous behavior if i called `context.log.foo` from user code was:

```       dagster_event = {"event_specific_data": [["AttributeError: 'DagsterLog' object has no attribute 'gargle'\n", ["  File \"/Users/max/dev/dagster/python_modules/dagster/dagster/core/errors.py\", line 177, in user_code_error_boundary\n    yield\n", "  File \"/Users/max/dev/dagster/python_modules/dagster/dagster/core/execution_plan/simple_engine.py\", line 374, in _iterate_step_outputs_within_boundary\n    for step_output in gen:\n", "  File \"/Users/max/dev/dagster/python_modules/dagster/dagster/core/execution_plan/transform.py\", line 88, in _execute_core_transform\n    for step_output in _yield_transform_results(transform_context, inputs):\n", "  File \"/Users/max/dev/dagster/python_modules/dagster/dagster/core/execution_plan/transform.py\", line 52, in _yield_transform_results\n    for result in gen:\n", "  File \"/Users/max/dev/dagster/python_modules/dagster/dagster/core/definitions/decorators.py\", line 310, in transform\n    result = fn(context, **kwargs)\n", "  File \"/Users/max/dev/dagster/python_modules/dagster/dagster_tests/core_tests/test_pipeline_errors.py\", line 283, in bad_logger\n    context.log.gargle('bargle')\n"], "AttributeError"]], "event_type_value": "STEP_FAILURE", "pipeline_name": "test_bad_log", "solid_definition_name": "bad_logger", "solid_name": "bad_logger", "step_key": "bad_logger.transform", "step_kind_value": "TRANSFORM", "tags": {"pipeline": "test_bad_log", "solid": "bad_logger", "solid_definition": "bad_logger", "step_key": "bad_logger.transform"}}```

File: python_modules/dagster/dagster/core/log.py (hunk 1 above)
Comment (mgasner) [reply python_modules/dagster/dagster/core/log.py]: ok, i think i might have a solution and will throw it up. it makes implementing a logger that defines new log levels a little trickier, but that's ok.

File: python_modules/dagster/dagster/core/log.py (hunk 3 above)
Comment (schrockn) [reply python_modules/dagster/dagster/core/log.py]: Right I guess what I'm saying is that I'm a bit concerned about the ability to inject errors into the pipeline for what looks like a totally valid configuration change. My instinct is that if you are provided a set of mixed loggers we should force the user to either 1) define some interface that maps a well known set of methods to the correct calls in the underlying loggers to 2) expose some explicit API that does the moral equivalent of "try to log to this specific logger but only if it exists"

**Pull request closed without merging**